Add _native nested_tensor_from_mask

zrphercule opened the pull request:
Summary: For user to convert nested tensor more easily. Some impl detail might change on user's need.

Test Plan: buck test mode/dev caffe2/test:nn -- test_nested_tensor_from_mask

Differential Revision: D36191182

cpuhrsch commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4686,6 +4686,10 @@
     CPU, NestedTensorCPU: transform_bias_rescale_qkv_cpu
     CUDA, NestedTensorCUDA: transform_bias_rescale_qkv_cuda
 
+- func: _nested_tensor_from_mask(Tensor t, Tensor mask, float min_padding_rate=0) -> Tensor
```
What is the idea behind "min_padding_rate"?

zrphercule replied:
The reason we have this design is when padding rate is pretty low, which means the tensor is almost not padded, using nested tensor will not have any perf benefit; Instead, it will damage the perf. So we would like to have this while setting it to 0 by default.

Since we actually have a diff calling this nested_tensor_from_mask by default in BetterTransformer, @erichan1 and I both feel it is might be better to keep this functionality. Users can have 0 knowledge of this min_padding_rate if they dont want to use it. Idea?

cpuhrsch commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -164,6 +164,32 @@ Tensor NestedTensor_gelu(const Tensor& self, c10::string_view approximate) {
 }
 
+Tensor NestedTensor_nested_tensor_from_mask(const Tensor& t, const Tensor& mask, double min_padding_rate) {
+    TORCH_CHECK(mask.scalar_type() == at::ScalarType::Bool, "Mask tensor should be Bool");
```
Say what you got instead. ```"Expected mask to be of ScalarType Bool, but got ", mask.scalar_type(), " instead."```

jbschlosser commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -164,6 +164,32 @@ Tensor NestedTensor_gelu(const Tensor& self, c10::string_view approximate) {
+    TORCH_CHECK(mask.scalar_type() == at::ScalarType::Bool, "Mask tensor should be Bool");
+    TORCH_CHECK(mask.dim() == 2, "Padding mask should be 2D");
+    TORCH_CHECK(t.dim() == 3, "Input should be a 3D tensor, N * L * D");
+    // One assumption is mask is always padded at the end , never in the middle
```
Can this assumption be validated efficiently for a nice error message if the mask is not structured correctly?

zrphercule replied:
I feel we can always validte it, but efficiently... Maybe no, it is always a O(mask.size()) solution?

jbschlosser commented on test/test_nn.py:
```diff
@@ -5434,6 +5434,38 @@ def test_pad_scalar_error(self):
         self.assertRaises(RuntimeError, lambda: F.pad(inputs, (1, 1)))
         self.assertRaises(RuntimeError, lambda: F.pad(inputs, (1,)))
 
+    def test_nested_tensor_from_mask(self):
+        B = 3
+        L = 7
+        D = 4
+        mask_size = [4, 7, 1]
+        i1 = torch.randn(4, D)
+        i2 = torch.randn(7, D)
+        i3 = torch.randn(1, D)
+
+        input = []
+        for i in [i1, i2, i3]:
+            masked = torch.zeros((L - i.size(0), D))
+            masked = torch.cat((i, masked))
+            input.append(masked.reshape(1, L, D))
+        input = torch.cat(input)
+
+        masks = []
+        for i in range(B):
+            m = []
+            for j in range(L):
+                if j < mask_size[i]:
+                    m.append(False)
+                else:
+                    m.append(True)
+            masks.append(torch.tensor(m).reshape(1, L))
+        mask = torch.cat(masks)
+
+        nt = torch._nested_tensor_from_mask(input, mask)
+        input_convert = nt.to_padded_tensor(0.)
+
+        self.assertEqual(input, input_convert)
```
This test logic seems a bit complex. Would something like this work?
```python
# L is the ragged dimension
N, L, D = 10, 7, 5
t = torch.randn(N, L, D)

# Build a random padding mask
mask = torch.ones(N, L, dtype=torch.bool)
for i in range(N):
    end = torch.randint(1, L, size=()).item()
    mask[i, :end] = False

nt = torch._nested_tensor_from_mask(t, mask)
...
```

zrphercule replied:
Yes, I actually have some comments before the testcase. I was prepare for a nested vs. nested comparision, but ends with finding out nested tensor does not have eq(). Will update this test.

jbschlosser commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -164,6 +164,32 @@ Tensor NestedTensor_gelu(const Tensor& self, c10::string_view approximate) {
+    // N * L
+    Tensor sizes = mask;
+    sizes.logical_not_();
+    auto real_padding_rate = 1 - (sizes.sum().item().toDouble() / sizes.numel());
+    if (real_padding_rate < min_padding_rate) {
+        return t;
+    }
```
As mentioned above: I suggest keeping the padding rate logic outside of this generic op.

jbschlosser commented on test/test_nn.py:
```diff
@@ -5434,6 +5434,38 @@ def test_pad_scalar_error(self):
         self.assertRaises(RuntimeError, lambda: F.pad(inputs, (1, 1)))
         self.assertRaises(RuntimeError, lambda: F.pad(inputs, (1,)))
 
+    def test_nested_tensor_from_mask(self):
```
Could possibly include tests validating that errors are thrown when shapes are mismatched

jbschlosser commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -164,6 +164,32 @@ Tensor NestedTensor_gelu(const Tensor& self, c10::string_view approximate) {
+    TORCH_CHECK(N == NN && L == LL, "Mask size should match input size");
+    // N * L
+    Tensor sizes = mask;
+    sizes.logical_not_();
```
this indicates to me that it's probably better to have `mask=True` be defined to mean "locations to include" rather than "locations to exclude"

zrphercule replied:
Yes, but in MHA and BetterTransformer, the current biggest customer of nested tensor, we are using such mask=False by defined, so let's keep it this way.

jbschlosser replied:
For MHA / BetterTransformer, the mask can be flipped there before passing it to this op, no? That would save an unnecessary flip for other users and improve performance.

`mask=True` for including also fits better with masking semantics used throughout the rest of PyTorch (e.g. [`masked_fill_`](https://pytorch.org/docs/stable/generated/torch.Tensor.masked_fill_.html), ['masked_scatter_'](https://pytorch.org/docs/stable/generated/torch.Tensor.masked_scatter_.html), etc.). It's unfortunate that MHA is an exception to this, but there have been complaints about it and ideally I think we shouldn't follow in its footsteps for a new, generic op.

mikekgfb replied:
What are those other users?  This is a private function, and was not intended as a new generic op.  I can see the argument of aligning with with other mask definitions, but the innermost loop elements are MHA and Layer, so that's where the most conversions would happen.  

Call it padding, not mask, and let conversion take either mask or padding - with padding being inverse polarity of mask?

mikekgfb replied:
Arguably what we have here is a description (mask) of padding, not the usual mask of what's "valid". At least call the inputs in the code prominantly padding and before inversion, and mask after inversion if we only support 1 polarity of the mask?

jbschlosser reviewed (commented):
I think an op like this is pretty useful in general, so I reviewed this from that perspective.

Given that, I think the opposite masking semantics would be clearer: `mask=True` locations are *included* in the nested tensor, leaving the concept of "padding" to applications that need it. But I won't die on this hill :)

jbschlosser commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -164,6 +164,36 @@ Tensor NestedTensor_gelu(const Tensor& self, c10::string_view approximate) {
+    sizes = sizes.cumsum(1).select(1, L - 1);
+    nums = nums.to(sizes.options());
+
+    TORCH_CHECK(sizes.equal(nums), "Mask should always padded at the end, never in the middle");
```
nit: for clarity, maybe `Mask must be left-aligned without gaps`

jbschlosser reviewed (approved):
Semantics LGTM, stamping. Minor wording comment for error msg to match new mask semantics but otherwise nice :)

zrphercule2 committed:
Add native nested_tensor_from_mask (#76942)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76942

For user to convert nested tensor more easily. Some impl detail might change on user's need.

Test Plan: buck test mode/dev caffe2/test:nn -- test_nested_tensor_from_mask

Reviewed By: mikekgfb

Differential Revision: D36191182

fbshipit-source-id: 1abe51155b6aa33a7f235225f324433601a866b7
